### PR TITLE
Player ranged attack

### PR DESCRIPTION
**Pull request by gracexuwt**
## ClickUp ticket link

[Player Long Range Attack](https://app.clickup.com/t/8685qfv3d)

## Implementation description
Prefab shooting is implemented for player ranged attacks.
Bullet implementation could be found in TemplateProjectile, and impact effect implementation could be found in Temporary Projectile Effect.
The bullet shoots from the direction where the player is facing by using the RangeFirePoint.

## Steps to test
1. Play game and press down on Ctrl button
2. Make sure that the bullet is shot after the player animation is finished, and that the bullet hits the target (either enemy or platform). If there is no obstacle in the direction of the player, then the bullet should "fly" offscreen
3. Make sure that the instantiated bullet and impact effect (the effect when a bullet hits a target) prefabs would be deleted from the hierarchy once the ranged attack is finished.
4. If enemy is hit, ensure that enemy health decreases

## What should reviewers focus on?
Ranged attack mechanics should run smoothly

## Checklist
- [x] My PR name is descriptive
- [x] My commit messages are descriptive. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR
- [x] My code is organized, and commented and documented so that another dev can understand what I did

### CODE REVIEW

File: bci-game/Assets/Scripts/Entity/Player/PlayerAttackRange.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        
+        private IEnumerator Attack(float duration)
+        {
+            yield return new WaitForSeconds(animationTime);
+            //shoot after animation finishes playing
+            Instantiate(rangeProjectile, firePoint.position, transform.localScale.x > 0 ? Quaternion.identity : Quaternion.Euler(0, 180, 0));
+            //attack cooldown
+            yield return new WaitForSeconds(duration);
+            canFire = true;
+        }
```
Comment (kimchiloof) [inline bci-game/Assets/Scripts/Entity/Player/PlayerAttackRange.cs:56]: Would this be better? Moves the sound effect to occur when the bullet spawns, separates shooting and cooldown more explicitly.

```c#
private IEnumerator Shoot()
{
    canFire = false;
    
    animator.SetTrigger(IsRanged);
    
    // shoot after animation finishes playing
    yield return new WaitForSeconds(animationTime);
    soundController.PlaySound(rangeAttackSounds, rangeAttackVolume);
    Instantiate(rangeProjectile, firePoint.position, transform.localScale.x > 0 ? Quaternion.identity : Quaternion.Euler(0, 180, 0));
    
    StartCoroutine(AttackCooldown(1f / rangeAttackSpeed));
}

private IEnumerator AttackCooldown(float duration)
{
    yield return new WaitForSeconds(duration);
    canFire = true;
}
```

File: bci-game/Assets/Scripts/Entity/Player/PlayerAttackRange.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            animator = GetComponent<Animator>();
+            soundController = GetComponent<CharacterSoundController>();
+        }
+
+        private void Update()
+        {
+            if (Input.GetButton("Fire1") && canFire)
+            {
+                Shoot();
```
Comment (kimchiloof) [inline bci-game/Assets/Scripts/Entity/Player/PlayerAttackRange.cs:35]: If making Shoot() to be a Coroutine:
```c#
StartCoroutine(Shoot());
```

**Review by kimchiloof — CHANGES_REQUESTED** [review]
Works well, delay fixed! :)

Just had a few thoughts on code readability, was wondering if these changes make it easier or harder to understand each function's purpose.

**Comment by gracexuwt** [comment]
Thanks for reviewing! Fixed code readability

**Review by kimchiloof — APPROVED** [review]

**Comment by kimchiloof** [comment]
Looks good to me! Will leave it up for a day or two if anyone else wants to have a look before it's merged, but otherwise all good to be merged into main.

**Pull request merged**